File: diskcount/sources/ebay.py
from __future__ import annotations

import base64
import time
from decimal import Decimal
from typing import Any

import httpx

from diskcount.domain import Deal
from diskcount.parsing import decimal_from_text, normalize_condition, normalize_media_type, parse_capacity_tb


class EbayBrowseSource:
    name = "ebay"

    def __init__(
        self,
        client_id: str,
        client_secret: str,
        search_queries: list[str],
        marketplace_id: str = "EBAY_FR",
        scope: str = "https://api.ebay.com/oauth/api_scope",
        limit: int = 50,
        category_ids: list[str] | None = None,
    ) -> None:
        self.client_id = client_id
        self.client_secret = client_secret
        self.search_queries = search_queries
        self.marketplace_id = marketplace_id
        self.scope = scope
        self.limit = limit
        self.category_ids = category_ids or []
        self._token: str | None = None
        self._token_expires_at = 0.0

    async def fetch(self, client: httpx.AsyncClient) -> list[Deal]:
        token = await self._get_token(client)
        deals: list[Deal] = []
        headers = {
            "Authorization": f"Bearer {token}",
            "X-EBAY-C-MARKETPLACE-ID": self.marketplace_id,
        }
        for query in self.search_queries:
            params: dict[str, str] = {"q": query, "limit": str(self.limit)}
            if self.category_ids:
                params["category_ids"] = ",".join(self.category_ids)
            response = await client.get(
                "https://api.ebay.com/buy/browse/v1/item_summary/search",
                headers=headers,
                params=params,
            )
            response.raise_for_status()
            deals.extend(parse_ebay_search_response(response.json(), source_name=self.name))
        return deals

    async def _get_token(self, client: httpx.AsyncClient) -> str:
        now = time.time()
        if self._token and now < self._token_expires_at - 60:
            return self._token
        credentials = base64.b64encode(f"{self.client_id}:{self.client_secret}".encode("utf-8")).decode("ascii")
        response = await client.post(
            "https://api.ebay.com/identity/v1/oauth2/token",
            headers={
                "Authorization": f"Basic {credentials}",
                "Content-Type": "application/x-www-form-urlencoded",
            },
            data={"grant_type": "client_credentials", "scope": self.scope},
        )
        response.raise_for_status()
        payload = response.json()
        self._token = str(payload["access_token"])
        self._token_expires_at = now + int(payload.get("expires_in", 7200))
        return self._token


def parse_ebay_search_response(payload: dict[str, Any], source_name: str = "ebay") -> list[Deal]:
    deals: list[Deal] = []
    for item in payload.get("itemSummaries", []):
        title = str(item.get("title") or "")
        subtitle = str(item.get("subtitle") or "")
        short_description = str(item.get("shortDescription") or "")
        haystack = f"{title} {subtitle} {short_description}"
        price = item.get("price") or item.get("currentBidPrice") or {}
        currency = str(price.get("currency") or "").upper()
        if currency and currency != "EUR":
            continue
        price_eur = decimal_from_text(str(price.get("value") or ""))
        capacity_tb = parse_capacity_tb(haystack)
        media_type = normalize_media_type(haystack)
        if price_eur is None or capacity_tb is None or capacity_tb <= 0 or media_type is None:
            continue
        condition = normalize_condition(str(item.get("condition") or ""))
        price_per_tb = (price_eur / capacity_tb).quantize(Decimal("0.01"))
        item_id = str(item.get("itemId") or item.get("legacyItemId") or "")
        url = str(item.get("itemWebUrl") or "")
        deals.append(
            Deal(
                source=source_name,
                external_id=item_id or url,
                title=title,
                url=url,
                price_eur=price_eur.quantize(Decimal("0.01")),
                price_per_tb=price_per_tb,
                capacity_tb=capacity_tb,
                condition=condition,
                media_type=media_type,
                raw={"marketplace": item.get("itemLocation"), "seller": item.get("seller")},
            )
        )
    return deals

File: internal/sources/ebay.go
package sources

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/Balrog57/DiskCount/internal/domain"
	"github.com/Balrog57/DiskCount/internal/parsing"
	"github.com/Balrog57/DiskCount/internal/scraper"
)

func init() {
	Register(func(r *Registry) Source {
		cfg := r.Config()
		if cfg.EbayClientID == "" || cfg.EbayClientSecret == "" || len(cfg.EbaySearchQueries) == 0 {
			return nil
		}
		return &Ebay{
			http:      r.HTTP(),
			clientID:  cfg.EbayClientID,
			secret:    cfg.EbayClientSecret,
			queries:   cfg.EbaySearchQueries,
			apiBase:   "https://api.ebay.com",
			oauthBase: "https://api.ebay.com/identity/v1/oauth2/token",
		}
	})
}

// Ebay implements the eBay Browse API source. It authenticates via the OAuth2
// client-credentials grant and searches for configured queries, mapping each
// item summary to a domain.Deal.
type Ebay struct {
	http      scraper.Fetcher
	clientID  string
	secret    string
	queries   []string
	apiBase   string
	oauthBase string

	mu          sync.Mutex
	token       string
	tokenExpiry time.Time
}

func (s *Ebay) Name() string { return "ebay" }

func (s *Ebay) Info() SourceInfo {
	return SourceInfo{
		Name:        "ebay",
		Description: "API eBay Browse (recherche d'articles par mots-cles)",
		Categories:  []string{"api"},
		Requires:    []string{"EBAY_CLIENT_ID", "EBAY_CLIENT_SECRET", "EBAY_SEARCH_QUERIES"},
		Version:     "1",
	}
}

func (s *Ebay) Fetch(ctx context.Context) ([]domain.Deal, error) {
	token, err := s.ensureToken(ctx)
	if err != nil {
		return nil, fmt.Errorf("ebay oauth: %w", err)
	}
	var deals []domain.Deal
	for _, q := range s.queries {
		items, err := s.search(ctx, token, q)
		if err != nil {
			slog.Warn("ebay search", "query", q, "err", err)
			continue
		}
		deals = append(deals, items...)
	}
	slog.Debug("ebay", "deals", len(deals))
	return deals, nil
}

// ensureToken returns a valid OAuth2 access token, refreshing it if necessary.
func (s *Ebay) ensureToken(ctx context.Context) (string, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.token != "" && time.Now().Before(s.tokenExpiry.Add(-60*time.Second)) {
		return s.token, nil
	}
	creds := base64.StdEncoding.EncodeToString([]byte(s.clientID + ":" + s.secret))
	body := "grant_type=client_credentials&scope=" + url.QueryEscape("https://api.ebay.com/oauth/api_scope")
	resp, err := s.http.PostWithHeaders(ctx, s.oauthBase, body, map[string]string{
		"Content-Type":  "application/x-www-form-urlencoded",
		"Authorization": "Basic " + creds,
	})
	if err != nil {
		return "", err
	}
	var tok struct {
		AccessToken string `json:"access_token"`
		ExpiresIn   int    `json:"expires_in"`
		TokenType   string `json:"token_type"`
	}
	if err := json.Unmarshal([]byte(resp), &tok); err != nil {
		return "", fmt.Errorf("parse token: %w", err)
	}
	if tok.AccessToken == "" {
		return "", fmt.Errorf("empty access token in response: %s", resp)
	}
	s.token = tok.AccessToken
	exp := 7200
	if tok.ExpiresIn > 0 {
		exp = tok.ExpiresIn
	}
	s.tokenExpiry = time.Now().Add(time.Duration(exp) * time.Second)
	slog.Debug("ebay token acquired", "expires_in", exp)
	return s.token, nil
}

func (s *Ebay) search(ctx context.Context, token, query string) ([]domain.Deal, error) {
	u := fmt.Sprintf("%s/buy/browse/v1/item_summary/search?q=%s&limit=50&filter=buyingOptions:{FIXED_PRICE}",
		s.apiBase, url.QueryEscape(query))
	resp, err := s.http.GetWithAuth(ctx, u, "Bearer "+token)
	if err != nil {
		return nil, err
	}
	var result struct {
		ItemSummaries []struct {
			ItemID        string `json:"itemId"`
			Title         string `json:"title"`
			ItemWebURL    string `json:"itemWebUrl"`
			Price         struct {
				Value    string `json:"value"`
				Currency string `json:"currency"`
			} `json:"price"`
			Condition struct {
				ConditionID string `json:"conditionId"`
				DisplayName string `json:"condition"`
			} `json:"condition"`
			Brand string `json:"brand"`
			Image struct {
				ImageURL string `json:"imageUrl"`
			} `json:"image"`
		} `json:"itemSummaries"`
	}
	if err := json.Unmarshal([]byte(resp), &result); err != nil {
		return nil, fmt.Errorf("parse search: %w", err)
	}

	var deals []domain.Deal
	for _, it := range result.ItemSummaries {
		if it.Title == "" || it.ItemWebURL == "" || it.Price.Value == "" {
			continue
		}
		price, err := parseFloatClean(it.Price.Value)
		if err != nil || price <= 0 {
			continue
		}
		tb, err := parsing.ParseCapacityTB(it.Title)
		if err != nil || tb <= 0 {
			continue // not a disk listing with parseable capacity
		}
		pt := price / tb
		cond := conditionFromEbay(it.Condition.ConditionID, it.Condition.DisplayName)
		classText := it.Title + " " + it.Brand
		media := normalMedia(classText)
		dc := parsing.NormalizeDriveCategory(classText, media)
		ifaces := parsing.NormalizeInterfaces(classText)
		var extID *string
		if it.ItemID != "" {
			extID = &it.ItemID
		}
		deals = append(deals, domain.Deal{
			Source: "ebay", Title: it.Title, URL: it.ItemWebURL,
			PriceEUR: round2(price), PricePerTB: round2(pt), CapacityTB: round2(tb),
			Condition: &cond, MediaType: media,
			ExternalID:    extID,
			DriveCategory: dc, Interfaces: ifaces,
			Brand: strPtr(it.Brand),
			ObservedAt: domain.UTCNow(),
		})
	}
	return deals, nil
}

// conditionFromEbay maps eBay condition IDs to our Condition type.
// eBay uses numeric condition IDs: 1000=New, 2000-3000=Refurbished/Used.
func conditionFromEbay(condID, displayName string) domain.Condition {
	id := strings.TrimSpace(condID)
	if strings.HasPrefix(id, "1") { // 1000 = New
		return domain.ConditionNew
	}
	lname := strings.ToLower(displayName)
	if strings.Contains(lname, "new") {
		return domain.ConditionNew
	}
	return domain.ConditionUsed
}
